nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,520 @@
+      <label for="firstName">First Name: </label>
+      <input type="text" name="firstName" required>
+      <label for="lastName">Last Name: </label>
+      <input type="text" name="lastName" required>
+      <button type="submit">Submit</button>
+    </form>
+    <a href="/users">View all users</a>
+  </body>
+</html>
+```
+
+```ejs
+<!-- views/users.ejs -->
+<!DOCTYPE html>
+<html>
+  <body>
+    <h1><%= title %></h1>
+    <ul>
+      <% users.forEach(user => { %>
+        <li>
+          ID: <%= user.userId %>, Name: <%= user.firstName %> <%= user.lastName %> 
+        </li>
+      <% }); %> 
+    </ul>
+    <a href="/">Back to Home</a>
+  </body>
+</html>
+```
+
+And we'll finally add the logic for this router in our controller:
+
+```javascript
+// controllers/usersController.js
+// User is a factory function which lets us model what we would see in a database.
+// We will explore this concept more as we dive deeper into databases.
+const User = (userId, firstName, lastName) => ({ userId, firstName, lastName });
+const asyncHandler = require("express-async-handler");
+
+let userStorage = {};
+let userId = 0;
+
+exports.users_list = asyncHandler(async (req, res) => {
+  res.render("users", { title: "User List", users: Object.values(userStorage) });
+});
+
+exports.users_create_post = asyncHandler(async (req, res) => {
+  const { firstName, lastName } = req.body;
+  userStorage[userId] = User(userId, firstName, lastName);
+  userId++;
+  res.redirect("/users");
+});
+```
+
+We can test our server with: `node --watch app.js`.
+
+You'll see we're able to add new users at `http://localhost:3000`, as well as list all existing users at `http://localhost:3000/users`.
+
+Let's add a few methods to our controller for validating and sanitizing our form to get the type of data we want.
+
+```javascript
+const {body, validationResult} = require("express-validator")
+
+const validate_user = [
+  body("firstName").trim().isAlpha().isLength({ min: 1, max: 10 }).escape(),
+  body("lastName").trim().isAlpha().isLength({ min: 1, max: 10 }).escape(),
+];
+
+// We can pass an entire array of middleware validations to our controller.
+exports.users_create_post = [
+  validate_user,
+  asyncHandler(async (req, res) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(400).render("index", { title: "Userbase", errors: errors.array() });
+    } else {
+      // Original logic
+    }
+  })
+];
+```
+
+And we need to update our `index.ejs` view to render these errors. Let's create a new partial: `errors.ejs`:
+
+```ejs
+<!-- views/partials/errors.ejs -->
+<% if (locals.errors) {%>
+  <ul>
+    <% errors.forEach(function(error) { %>
+    <li><%= error.msg %></li>
+    <% }); %>
+  </ul>
+<% } %>
+```
+
+And we'll include the partial just above our form in `index.ejs`.
+
+```ejs
+<%- include("partials/errors.ejs") %>
+```
+
+If the form is filled out incorrectly, you'll see a page containing all the errors.
```

Comment:
> you'll see a page containing all the errors.

won't the errors show up on the existing form page?

Reply:
Yeah that was just clumsy writing on my part. It should be `the page contains`.